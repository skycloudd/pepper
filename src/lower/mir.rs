#![allow(dead_code)]

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub struct Mir {
    pub functions: Vec<Function>,
}

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub struct Function {
    pub name: Name,
    pub params: Vec<FuncParam>,
    pub return_ty: Type<Primitive>,
    pub body: TypedExpression,
}

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub struct FuncParam {
    pub name: Name,
    pub ty: Type<Primitive>,
}

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub struct TypedExpression {
    pub expr: Expression,
    pub ty: Type<Primitive>,
}

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub enum Expression {
    Number(f64),
    Bool(bool),
    Variable(Name),
    Intrinsic(Box<Intrinsic>),
    Call {
        callee: Box<TypedExpression>,
        args: Vec<TypedExpression>,
    },
    Match {
        expr: Box<TypedExpression>,
        arms: Vec<MatchArm>,
    },
}

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub struct MatchArm {
    pub pattern: Pattern,
    pub body: TypedExpression,
}

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub struct Pattern {
    pub pattern_type: PatternType,
    pub condition: Option<TypedExpression>,
}

#[derive(Clone, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub enum PatternType {
    Wildcard,
    Variable(Name),
    Number(f64),
    Bool(bool),
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub enum Type<P> {
    Error,
    Primitive(P),
    Tuple(Vec<Self>),
    Never,
    Function {
        params: Vec<Self>,
        return_ty: Box<Self>,
    },
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub enum Primitive {
    Number,
    Bool,
}

#[derive(Clone, Copy, Debug)]
#[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
pub struct Name(u32);

impl Name {
    pub const fn new(id: u32) -> Self {
        Self(id)
    }
}

bin_op_intrinsics! {
    AddNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    SubNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    MulNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    DivNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    EqNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    NeqNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    LtNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    LteNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    GtNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    GteNumbers { lhs, rhs }, Type::Primitive(Primitive::Number),
    NegNumber { expr }, Type::Primitive(Primitive::Number),
    EqBools { lhs, rhs }, Type::Primitive(Primitive::Bool),
    NeqBools { lhs, rhs }, Type::Primitive(Primitive::Bool),
    AndBools { lhs, rhs }, Type::Primitive(Primitive::Bool),
    OrBools { lhs, rhs }, Type::Primitive(Primitive::Bool),
    NotBool { expr }, Type::Primitive(Primitive::Bool),
}

macro_rules! bin_op_intrinsics {
    (
        $(
            $name:ident { $($field:ident),* $(,)? },
            $sub_expr_ty:expr,
        )* $(,)?
    ) => {
        #[derive(Clone, Debug)]
        #[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
        pub enum Intrinsic {
            $(
                $name($name),
            )*
        }

        $(
            #[derive(Clone, Debug)]
            #[cfg_attr(test, derive(serde::Deserialize, serde::Serialize))]
            pub struct $name {
                $(pub $field: TypedExpression,)*
            }

            impl $name {
                pub fn new($($field: TypedExpression,)*) -> Self {
                    $(
                        assert_eq!($field.ty, $sub_expr_ty);
                    )*

                    Self { $($field,)* }
                }
            }
        )*

        paste::paste! {
            impl Intrinsic {
                $(
                    pub fn [<$name:snake>]($($field: TypedExpression,)*) -> Self {
                        Self::$name($name::new($($field,)*))
                    }
                )*
            }
        }
    };
}
use bin_op_intrinsics;
