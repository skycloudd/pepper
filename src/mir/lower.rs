use super::{
    BinaryOp, Expression, Function, FunctionParam, FunctionType, Identifier, Mir, Primitive, Type,
    TypedExpression, UnaryOp,
};
use crate::{
    parser::ast,
    typecheck::typed_ast::{self, TypedAst},
};
use nonempty::NonEmpty;

pub fn lower(typed_ast: TypedAst) -> Mir {
    let mut functions = vec![];
    let mut main = vec![];

    for toplevel in typed_ast.toplevels {
        match toplevel.0 {
            typed_ast::TopLevel::Function(function) => functions.push(lower_function(function.0)),
            typed_ast::TopLevel::Do(expr) => main.push(lower_expression(expr.0)),
        }
    }

    Mir { functions, main }
}

fn lower_function(function: typed_ast::Function) -> Function {
    let name = lower_ident(function.name.0);

    let params = NonEmpty::collect(function.params.0.into_iter().map(|param| {
        let name = lower_ident(param.0.name.0);
        let ty = lower_type(param.0.ty.0);

        FunctionParam { name, ty }
    }))
    .unwrap();

    let return_ty = lower_type(function.return_ty.0);

    let body = lower_expression(function.body.0);

    Function {
        name,
        params,
        return_ty,
        body,
    }
}

const fn lower_ident(ident: ast::Identifier) -> Identifier {
    Identifier(ident.0 .0)
}

fn lower_type(ty: ast::Type<typed_ast::Primitive>) -> Type<Primitive> {
    match ty {
        ast::Type::Primitive(primitive) => Type::Primitive(lower_primitive(primitive)),
        ast::Type::Unit => Type::Unit,
        ast::Type::Never => Type::Never,
        ast::Type::Function(function_type) => Type::Function(lower_function_type(function_type)),
    }
}

const fn lower_primitive(primitive: typed_ast::Primitive) -> Primitive {
    match primitive {
        typed_ast::Primitive::Number => Primitive::Number,
        typed_ast::Primitive::Bool => Primitive::Bool,
    }
}

fn lower_function_type(
    function_type: ast::FunctionType<typed_ast::Primitive>,
) -> FunctionType<Primitive> {
    let params = function_type
        .params
        .0
        .into_iter()
        .map(|param| lower_type(param.0))
        .collect();

    let return_ty = lower_type(*function_type.return_ty.0);

    FunctionType {
        params,
        return_ty: Box::new(return_ty),
    }
}

fn lower_expression(expr: typed_ast::TypedExpression) -> TypedExpression {
    let ty = lower_type(expr.ty);

    TypedExpression {
        ty,
        expr: match expr.expr {
            typed_ast::Expression::Unit => Expression::Unit,
            typed_ast::Expression::Number(value) => Expression::Number(value.0),
            typed_ast::Expression::Bool(value) => Expression::Bool(value.0),
            typed_ast::Expression::Variable(name) => Expression::Variable(lower_ident(name.0)),
            typed_ast::Expression::BinaryOp { op, lhs, rhs } => Expression::BinaryOp {
                op: lower_binary_op(op.0),
                lhs: Box::new(lower_expression(*lhs.0)),
                rhs: Box::new(lower_expression(*rhs.0)),
            },
            typed_ast::Expression::UnaryOp { op, expr } => Expression::UnaryOp {
                op: lower_unary_op(op.0),
                expr: Box::new(lower_expression(*expr.0)),
            },
            typed_ast::Expression::Call { name, args } => Expression::Call {
                name: Box::new(lower_expression(*name.0)),
                args: NonEmpty::collect(
                    args.0
                        .into_iter()
                        .map(|arg| Box::new(lower_expression(*arg.0))),
                )
                .unwrap(),
            },
        },
    }
}

const fn lower_binary_op(op: ast::BinaryOp) -> BinaryOp {
    match op {
        ast::BinaryOp::Add => BinaryOp::Add,
        ast::BinaryOp::Sub => BinaryOp::Sub,
        ast::BinaryOp::Mul => BinaryOp::Mul,
        ast::BinaryOp::Div => BinaryOp::Div,
    }
}

const fn lower_unary_op(op: ast::UnaryOp) -> UnaryOp {
    match op {
        ast::UnaryOp::Neg => UnaryOp::Neg,
    }
}
