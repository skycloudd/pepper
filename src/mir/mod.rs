use crate::RODEO;
use lasso::Spur;
use malachite::Rational;
use nonempty::NonEmpty;

pub mod lower;

#[derive(Clone, Debug)]
pub struct Mir {
    pub functions: Vec<Function>,
    pub main: Vec<TypedExpression>,
}

#[derive(Clone, Debug)]
pub struct Function {
    pub name: Identifier,
    pub params: NonEmpty<FunctionParam>,
    pub return_ty: Type<Primitive>,
    pub body: TypedExpression,
}

#[derive(Clone, Debug)]
pub struct FunctionParam {
    pub name: Identifier,
    pub ty: Type<Primitive>,
}

#[derive(Clone, Debug)]
pub struct TypedExpression {
    pub expr: Expression,
    pub ty: Type<Primitive>,
}

#[derive(Clone, Debug)]
pub enum Expression {
    Unit,
    Number(Rational),
    Bool(bool),
    Variable(Identifier),
    BinaryOp {
        op: BinaryOp,
        lhs: Box<TypedExpression>,
        rhs: Box<TypedExpression>,
    },
    UnaryOp {
        op: UnaryOp,
        expr: Box<TypedExpression>,
    },
    Call {
        name: Box<TypedExpression>,
        args: NonEmpty<Box<TypedExpression>>,
    },
}

#[derive(Clone, Copy, Debug)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
}

#[derive(Clone, Copy, Debug)]
pub enum UnaryOp {
    Neg,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Type<P> {
    Primitive(P),
    Unit,
    Never,
    Function(FunctionType<P>),
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Primitive {
    Number,
    Bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FunctionType<P> {
    pub params: Vec<Type<P>>,
    pub return_ty: Box<Type<P>>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Identifier(pub Spur);

impl Identifier {
    pub fn resolve(self) -> &'static str {
        RODEO.resolve(&self.0)
    }
}
