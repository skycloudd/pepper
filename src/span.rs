use core::num::NonZeroUsize;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Span {
    start: usize,
    end: usize,
    file_id: FileId,
}

impl Span {
    pub fn zero(file_id: FileId) -> Self {
        chumsky::span::Span::new(file_id, 0..0)
    }

    pub const fn range(&self) -> core::ops::Range<usize> {
        self.start..self.end
    }
}

impl chumsky::span::Span for Span {
    type Context = FileId;

    type Offset = usize;

    fn new(context: Self::Context, range: core::ops::Range<Self::Offset>) -> Self {
        Self {
            start: range.start,
            end: range.end,
            file_id: context,
        }
    }

    fn context(&self) -> Self::Context {
        self.file_id
    }

    fn start(&self) -> Self::Offset {
        self.start
    }

    fn end(&self) -> Self::Offset {
        self.end
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct FileId(NonZeroUsize);

impl FileId {
    pub fn new(id: usize) -> Option<Self> {
        Some(Self(NonZeroUsize::new(id.checked_add(1)?)?))
    }

    pub fn get(self) -> usize {
        usize::from(self.0) - 1
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Spanned<T>(pub T, Span);

impl<T> Spanned<T> {
    pub const fn new(value: T, span: Span) -> Self {
        Self(value, span)
    }

    pub const fn span(&self) -> Span {
        self.1
    }

    pub fn boxed(self) -> Spanned<Box<T>> {
        Spanned::new(Box::new(self.0), self.1)
    }

    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> Spanned<U> {
        Spanned::new(f(self.0), self.1)
    }

    pub fn map_with_span<U>(self, f: impl FnOnce(T, Span) -> U) -> Spanned<U> {
        Spanned::new(f(self.0, self.1), self.1)
    }

    pub fn map_span(self, f: impl FnOnce(Span) -> Span) -> Self {
        Self(self.0, f(self.1))
    }

    pub const fn as_ref(&self) -> Spanned<&T> {
        Spanned::new(&self.0, self.1)
    }
}

impl<T, E> Spanned<Result<T, E>> {
    pub fn transpose(self) -> Result<Spanned<T>, E> {
        self.0.map(|value| Spanned::new(value, self.1))
    }
}

impl<T> Spanned<Box<T>> {
    pub fn unbox(self) -> Spanned<T> {
        Spanned::new(*self.0, self.1)
    }
}

impl<T: Clone> Spanned<&T> {
    pub fn cloned(self) -> Spanned<T> {
        self.map(Clone::clone)
    }
}

impl<T> core::ops::Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> core::ops::DerefMut for Spanned<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[cfg(test)]
impl<T: serde::Serialize> serde::Serialize for Spanned<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.0.serialize(serializer)
    }
}
