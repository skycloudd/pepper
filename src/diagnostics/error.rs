use super::{Diag, ErrorSpan};
use crate::span::Span;
use chumsky::error::{Rich, RichReason};
use codespan_reporting::diagnostic::Severity;
use core::fmt::Display;
use owo_colors::OwoColorize;
use std::borrow::Cow;

#[derive(Clone, Debug)]
pub enum Error {
    ExpectedFound {
        expected: Vec<String>,
        found: Option<String>,
        span: Span,
    },
    Custom {
        message: String,
        span: Span,
    },
    AmbiguousModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
    MissingModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
}

impl Diag for Error {
    fn message(&self) -> Cow<str> {
        match self {
            Self::ExpectedFound {
                expected,
                found,
                span: _,
            } => if expected.is_empty() {
                "Unexpected end of file".to_string()
            } else if expected.len() == 1 {
                format!(
                    "Expected {}, but found {}",
                    expected[0].green(),
                    found.as_deref().unwrap_or("something else")
                )
            } else {
                format!(
                    "Expected one of {}, but found {}",
                    expected
                        .iter()
                        .map(|expected| expected.green().to_string())
                        .collect::<Vec<_>>()
                        .join(", "),
                    found.as_deref().unwrap_or("something else")
                )
            }
            .into(),
            Self::Custom { message, span: _ } => message.into(),
            Self::AmbiguousModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Ambiguous module name".into(),
            Self::MissingModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Missing module".into(),
        }
    }

    fn spans(&self) -> Vec<ErrorSpan> {
        match self {
            Self::ExpectedFound {
                expected: _,
                found,
                span,
            } => found.as_ref().map_or_else(
                || vec![ErrorSpan::primary(*span)],
                |found| vec![ErrorSpan::primary_message(format!("Found {found}"), *span)],
            ),
            Self::Custom { message: _, span } => vec![ErrorSpan::primary(*span)],
            Self::AmbiguousModule {
                module_name: _,
                module_name_span,
                filenames: _,
            }
            | Self::MissingModule {
                module_name: _,
                module_name_span,
                filenames: _,
            } => vec![ErrorSpan::primary_message(
                "Declared here",
                *module_name_span,
            )],
        }
    }

    fn notes(&self) -> Vec<String> {
        match self {
            Self::AmbiguousModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "There exist multiple possible files for the module `{module_name}`:"
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::MissingModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "Could not find a file for the module `{module_name}`"
            ))
            .chain(core::iter::once(
                "These are the possible files for this module:".into(),
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::ExpectedFound { .. } | Self::Custom { .. } => {
                vec![]
            }
        }
    }

    fn kind(&self) -> Severity {
        Severity::Error
    }
}

pub fn convert(error: &Rich<impl Display, Span, &str>) -> Vec<Error> {
    fn convert_inner(reason: &RichReason<impl Display, &str>, span: Span) -> Vec<Error> {
        match reason {
            RichReason::ExpectedFound { expected, found } => vec![Error::ExpectedFound {
                expected: expected.iter().map(ToString::to_string).collect(),
                found: found.as_ref().map(|f| f.to_string()),
                span,
            }],
            RichReason::Custom(message) => vec![Error::Custom {
                message: message.to_owned(),
                span,
            }],
            RichReason::Many(reasons) => reasons
                .iter()
                .flat_map(|r| convert_inner(r, span))
                .collect(),
        }
    }

    convert_inner(error.reason(), *error.span())
}
