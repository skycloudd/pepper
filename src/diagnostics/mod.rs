use crate::span::Span;
use codespan_reporting::diagnostic::{LabelStyle, Severity};
use std::borrow::Cow;

pub mod error;
pub mod report;

pub trait Diag {
    fn message(&self) -> Cow<str>;
    fn spans(&self) -> Vec<ErrorSpan>;
    fn notes(&self) -> Vec<String>;
    fn kind(&self) -> Severity;
}

#[derive(Debug)]
pub struct ErrorSpan {
    message: Option<String>,
    span: Span,
    label_style: LabelStyle,
}

impl ErrorSpan {
    pub const fn primary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub fn primary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub const fn secondary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Secondary,
        }
    }

    pub fn secondary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Secondary,
        }
    }
}
