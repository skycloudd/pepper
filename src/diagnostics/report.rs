use super::{Diag, ErrorSpan};
use chumsky::span::Span as _;
use codespan_reporting::diagnostic::{Diagnostic, Label};

pub fn report(diagnostic: &impl Diag) -> Diagnostic<usize> {
    codespan_reporting::diagnostic::Diagnostic::<usize>::new(diagnostic.kind())
        .with_message(diagnostic.message())
        .with_labels(
            diagnostic
                .spans()
                .into_iter()
                .map(|span| match span {
                    ErrorSpan::Primary(message, span) => {
                        let mut label =
                            Label::primary(span.context().id(), span.start()..span.end());

                        if let Some(message) = message {
                            label = label.with_message(message);
                        }

                        label
                    }
                    ErrorSpan::Secondary(message, span) => {
                        let mut label =
                            Label::secondary(span.context().id(), span.start()..span.end());

                        if let Some(message) = message {
                            label = label.with_message(message);
                        }

                        label
                    }
                })
                .collect(),
        )
        .with_notes(diagnostic.notes())
}
