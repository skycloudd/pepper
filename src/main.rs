use camino::{Utf8Path, Utf8PathBuf};
use chumsky::{input::Input as _, span::Span as _, Parser as _};
use clap::Parser;
use codespan_reporting::{
    files::SimpleFiles,
    term::{
        self,
        termcolor::{ColorChoice, StandardStream},
    },
};
use diagnostics::{
    error::{convert, Error},
    report::report,
};
use lasso::ThreadedRodeo;
use lower::mir::Mir;
use span::{FileId, Span};
use std::{fs::read_to_string, process::ExitCode, sync::LazyLock};

pub mod diagnostics;
mod lexer;
mod lower;
mod parser;
pub mod scopes;
pub mod span;
mod typecheck;

static RODEO: LazyLock<ThreadedRodeo> = LazyLock::new(ThreadedRodeo::new);

#[derive(Debug, Parser)]
struct Args {
    filename: Utf8PathBuf,
}

fn main() -> ExitCode {
    let args = Args::parse();

    let mut files = SimpleFiles::new();

    match run_mir(&args, &mut files) {
        Ok(mir) => {
            println!("{mir:#?}");
            ExitCode::SUCCESS
        }
        Err(errors) => {
            let writer = StandardStream::stderr(ColorChoice::Auto);
            let term_config = term::Config::default();

            for error in &errors {
                let diagnostic = report(error);

                term::emit(&mut writer.lock(), &term_config, &files, &diagnostic).unwrap();
            }

            ExitCode::FAILURE
        }
    }
}

fn run_mir<'path>(
    args: &'path Args,
    files: &mut SimpleFiles<&'path Utf8Path, String>,
) -> Result<Mir, Vec<Error>> {
    if !args.filename.is_file() {
        return Err(vec![Error::Ice(format!(
            "'{}' is not a file",
            args.filename
        ))]);
    }

    let source =
        read_to_string(&args.filename).map_err(|error| vec![Error::Ice(error.to_string())])?;

    let file_id = FileId::new(files.add(&args.filename, source.to_string()));

    let mut errors = vec![];

    let (tokens, lexer_errors) = lexer::lexer()
        .parse(source.with_context(file_id))
        .into_output_errors();

    errors.extend(lexer_errors.iter().flat_map(|error| convert(error)));

    let (ast, parser_errors) = tokens.as_ref().map_or_else(
        || (None, vec![]),
        |tokens| {
            let eoi = tokens
                .last()
                .map_or_else(|| Span::zero(file_id), |(_, span)| span.to_end());

            parser::parser()
                .parse(tokens.spanned(eoi))
                .into_output_errors()
        },
    );

    errors.extend(parser_errors.iter().flat_map(|error| convert(error)));

    let (typed_ast, typecheck_errors) = ast.map_or_else(
        || (None, vec![]),
        |ast| {
            let (typed_ast, errors) = typecheck::typecheck(ast);
            (Some(typed_ast), errors)
        },
    );

    errors.extend(typecheck_errors);

    if errors.is_empty() {
        let mir = lower::lower(typed_ast.unwrap());

        Ok(mir)
    } else {
        Err(errors)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mir_tests() {
        for file in std::fs::read_dir("tests/mir").unwrap() {
            let file = file.unwrap();

            let args = Args {
                filename: file.path().try_into().unwrap(),
            };

            let mir = run_mir(&args, &mut SimpleFiles::new())
                .map_err(|err| (args, err))
                .unwrap();

            insta::with_settings!({
                description => std::fs::read_to_string(file.path()).unwrap().trim(),
                info => &file.path().to_string_lossy(),
                snapshot_suffix => file.path().file_stem().unwrap().to_string_lossy(),
            }, {
                insta::assert_yaml_snapshot!(mir);
            });
        }
    }
}
