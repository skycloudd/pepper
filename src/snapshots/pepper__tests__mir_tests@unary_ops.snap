---
source: src/main.rs
description: "func main() -> int = 0\nfunc neg(a: int) -> int = -a\nfunc negf(a: float) -> float = -a\nfunc not(a: bool) -> bool = !a"
expression: mir
info: tests/mir/unary_ops.pr
snapshot_kind: text
---
functions:
  - name: 0
    params: []
    return_ty:
      Primitive: Int
    body:
      expr:
        Int: 0
      ty:
        Primitive: Int
  - name: 1
    params:
      - name: 4
        ty:
          Primitive: Int
    return_ty:
      Primitive: Int
    body:
      expr:
        Intrinsic:
          NegInt:
            expr:
              expr:
                Variable: 4
              ty:
                Primitive: Int
      ty:
        Primitive: Int
  - name: 2
    params:
      - name: 5
        ty:
          Primitive: Float
    return_ty:
      Primitive: Float
    body:
      expr:
        Intrinsic:
          NegFloat:
            expr:
              expr:
                Variable: 5
              ty:
                Primitive: Float
      ty:
        Primitive: Float
  - name: 3
    params:
      - name: 6
        ty:
          Primitive: Bool
    return_ty:
      Primitive: Bool
    body:
      expr:
        Intrinsic:
          NotBool:
            expr:
              expr:
                Variable: 6
              ty:
                Primitive: Bool
      ty:
        Primitive: Bool
externs: []
